### DISCUSSION TITLE
Continuous message processing

### DISCUSSION THREAD

**Discussion by sourcebox** [emilk/egui#995] [Q&A]
I need some functionality to continuously process messages, even when the application is in the background. Currently I do this in the `update()` function:
```rust
ctx.request_repaint();

while let Ok(message) = self.message_channel.1.try_recv() {
    self.process_message(&message);
}
```
What I don't like here is the `ctx.request_repaint();` because this forces `update()` to run over and over, even when there's nothing to do GUI-wise and results in some unnecessary load. Without the repaint request, message processing will stop when there's no GUI action, so omitting it is no option.

Any ideas for a better solution?

**Comment by coderedart** [comment]
why not just offload the processing part to a different thread?

**Reply by sourcebox** [reply]
> why not just offload the processing part to a different thread?

The processing part is already in another thread. This is where the messages are generated and sent to the main thread with the GUI. The GUI must then process these messages to update its state.

**Reply by coderedart** [reply]
ah.. looked around the docs, I don't think its possible. we cannot manually trigger a `update` without `ctx.request_repaint()`. even tried sending ctx to background thread to call `request_repaint()` after sleeping for a second, but it seems that won't work because eframe/epi doesn't check for repaint if there's no window events at all :(

**Comment by emilk** [comment]
Sounds like you want to use https://docs.rs/epi/0.15.0/epi/struct.Frame.html#method.repaint_signal to wake up the GUI to repaint only when needed

**Reply by sourcebox** [reply]
What I really want is an App::idle() method that is always called in a loop, even when there is nothing to update in the GUI. Otherwise, I have to pass CtxRef to other threads additionally, which I don't like because these processing threads are not even aware of anything GUI-related. They just send generic progress events.

Another issue I had in the past is the requirement for continuous update in conjunction with `rfd` on macOS. Without it, the file dialog got stuck until some mouse movement etc. was performed.

**Reply by correabuscar** [reply]
> Sounds like you want to use https://docs.rs/epi/0.15.0/epi/struct.Frame.html#method.repaint_signal to wake up the GUI to repaint only when needed

`You can also set up egui to only repaint when there is interaction (e.g. mouse movement).`
Can we have an example of this please?
I read it in the README under [CPU Usage](https://github.com/emilk/egui#cpu-usage).

Because even https://github.com/MrTanoshii/rusty-autoclicker has high enough CPU usage (2-4%, on 12 cores CPU) constantly, and it spins up CPU fans needlessly(compared to when it's not running it at all).

But more importantly, I'd like to make something with egui, but the high CPU usage in the demo is a huge deterrent!

**Comment by mchaliadzinau** [comment]
For MacOS even this workaround does not work:
> 2023-07-10 22:10:06.042 egui[76937:325250738] -[WinitWindow sender:]: unrecognized selector sent to instance 0x133104380

EDIT №1: 
It does not work for minimised window when I try to `request_user_attention` or set `frame.set_minimized(false)`, 
but it is possible to use another "workaround" by combination of `frame.set_visible(true);` and  `frame.focus();`.

EDIT №2:
Interesting... when I hide window from docker (task bar in macos), workaround I've mentioned above still works.
I'm confused why `minimised` does not works then.

**Comment by MCOfficer** [comment]
If I understand correctly, what we really want is a hook into the event loop that doesn't also trigger a redraw (with all the bells and whistles that entails). Unfortunately `eframe` doesn't provide that. Anyways, two options:

- it seems possible by switching out `eframe` for your own window setup & event loop, like https://github.com/emilk/egui/blob/master/crates/egui_glow/examples/pure_glow.rs . But honestly, that's too much effort in my book.. Maybe something like `egui_bevy` would work, though.
- This is a hack, but it gets us halfway there:

```rs
fn update(&mut self, ctx: &egui::Context, _frame: &mut eframe::Frame) {

	// If you want to be even more frugal, you can check `ViewportInfo.focused` instead
	let minimized = ctx.input(|i| i.viewport().minimized).unwrap_or_default(); 

	if !minimized {
		// Actual drawing code here, won't run if window is minimized
	}
	// Receive your events here, will always run

	ctx.request_repaint_after(Duration::from_millis(50));
}
```
